libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,722 @@
+		break; \
+	default: throwBadInstruction(); \
+	} \
+}
+#define ADD( x1, x2) ((x1) + (x2))
+#define MUL( x1, x2) ((x1) * (x2))
+#define SUB( x1, x2) ((x1) - (x2))
+#define DIV( x1, x2) ((x1) / (x2))
+#define MOD( x1, x2) ((x1) % (x2))
+#define LT(  x1, x2) ((x1) < (x2))
+#define GT(  x1, x2) ((x1) > (x2))
+#define EQ(  x1, x2) ((x1) == (x2))
+#define ZERO(x1, x2) ((x1) == 0)
+#define AND( x1, x2) ((x1) & (x2))
+#define OR(  x1, x2) ((x1) | (x2))
+#define XOR( x1, x2) ((x1) ^ (x2))
+#define NOT( x1, x2) (~x1)
+#define SHR( x1, x2) ((x1) >> (x2))
+#define SHL( x1, x2) ((x1) << (x2))
+#define ROL( x1, x2) (((x1) << x2)|((x1) >> (sizeof(x1) * 8 - (x2))))
+#define ROR( x1, x2) (((x1) >> x2)|((x1) << (sizeof(x1) * 8 - (x2))))
+
+void VM::xadd (uint8_t _b) { EVALXOPU(ADD, _b); }
+void VM::xmul (uint8_t _b) { EVALXOPU(MUL, _b); }
+void VM::xsub (uint8_t _b) { EVALXOPU(SUB, _b); }
+void VM::xdiv (uint8_t _b) { EVALXOPU(DIV, _b); }
+void VM::xsdiv(uint8_t _b) { EVALXOPS(DIV, _b); }
+void VM::xmod (uint8_t _b) { EVALXOPU(MOD, _b); }
+void VM::xsmod(uint8_t _b) { EVALXOPS(MOD, _b); }
+void VM::xlt  (uint8_t _b) { EVALXOPU(LT,  _b); }
+void VM::xslt (uint8_t _b) { EVALXOPS(LT,  _b); }
+void VM::xgt  (uint8_t _b) { EVALXOPU(GT,  _b); }
+void VM::xsgt (uint8_t _b) { EVALXOPS(GT,  _b); }
+void VM::xeq  (uint8_t _b) { EVALXOPU(EQ,  _b); }
+void VM::xzero(uint8_t _b) { EVALXOPU(ZERO,_b); }
+void VM::xand (uint8_t _b) { EVALXOPU(AND, _b); }
+void VM::xoor (uint8_t _b) { EVALXOPU(OR,  _b); }
+void VM::xxor (uint8_t _b) { EVALXOPU(XOR, _b); }
+void VM::xnot (uint8_t _b) { EVALXOPU(NOT, _b); }
+void VM::xshr (uint8_t _b) { EVALXOPU(SHR, _b); }
+void VM::xsar (uint8_t _b) { EVALXOPS(SHR, _b); }
+void VM::xshl (uint8_t _b) { EVALXOPU(SHL, _b); }
+void VM::xrol (uint8_t _b) { EVALXOPU(ROL, _b); }
+void VM::xror (uint8_t _b) { EVALXOPU(ROR, _b); }
+
+inline uint8_t pow2N(uint8_t n)
+{
+	static uint8_t exp[6] = { 1, 2, 4, 8, 16, 32 };
+	return exp[n];
+}
+
+inline uint8_t nElem(uint8_t _b)
+{
+	return pow2N((_b) & 0xf);
+}
+
+inline uint8_t elemType(uint8_t _b)
+{
+	return (_b) >> 4;
+}
+
+u256 VM::vtow(uint8_t _b, const u256& _in)
+{
+	u256 out;
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 8;
+			out |= v8x32(_in) [i];
+		}
+		break;
+	case 1:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 16;
+			out |= v16x16(_in)[i];
+		}
+		break;
+	case 2:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 32;
+			out |= v32x8(_in) [i];
+		}
+		break;
+	case 3:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			out << 64;
+			out |= v64x4(_in) [i];
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+	return out;
+}
+
+u256 VM::wtov(uint8_t _b, const u256& _in)
+{
+	u256 out;
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v8x32(out) [i] |= (uint8_t )(_in & 0xff);
```

Comment:
I think you could use simple assignment here instead of `|=` ?
You access different output vector elements on each iteration and they are all initially 0